doc/user/content/transform-data/patterns/percentiles.md

```diff
@@ -10,134 +10,306 @@ menu:
+   90-th percentile.
+
+   ```mzsql
+   SELECT bucket AS percentile90
+   FROM distribution
+   WHERE cumulative_distribution >= 0.9
+   ORDER BY cumulative_distribution
+   LIMIT 1;
+   ```
+
+1. To keep the `distribution` view up to date as new data comes into `input`,
+   you can create an index on the view. To optimize for percentile queries,
+   index on the `cumulative_distribution` column to support [point lookup on
+   the index](/concepts/indexes/#point-lookups):
+
+   ```mzsql
+   CREATE INDEX distribution_idx ON distribution (cumulative_distribution);
+   ```
 
-```mzsql
-CREATE VIEW histogram AS
-SELECT
-  value AS bucket,
-  count(*) AS frequency
-FROM input
-GROUP BY value;
-```
-
-To query percentiles from the view `histogram`, it's no longer possible to just order the `values` and pick a value from the right spot. Instead, the distribution of values needs to be reconstructed. It is done by determining the cumulative count (the sum of counts up through each bucket that came before) for each `bucket`. This is accomplished through a cross-join in the following view:
+## Using HDR histograms to compute approximate percentiles
 
-```mzsql
-CREATE VIEW distribution AS
-SELECT
-  h.bucket,
-  h.frequency,
-  sum(g.frequency) AS cumulative_frequency,
-  sum(g.frequency) / (SELECT sum(frequency) FROM histogram) AS cumulative_distribution
-FROM histogram g, histogram h
-WHERE g.bucket <= h.bucket
-GROUP BY h.bucket, h.frequency
-ORDER BY cumulative_distribution;
-```
+[HDR histograms](https://github.com/HdrHistogram/HdrHistogram) can be used to approximate percentiles in a space efficient manner that scales well even for large domains with many distinct values. HDR histograms reduce the precision of values that are tracked and use buckets with variable width. Buckets that are closer to 0 are smaller whereas buckets far away from 0 are wider. This works particularly well for data that exhibits a long tail of large values, e.g., latency measurements.
 
-The cumulative count and the cumulative distribution can then be used to query for arbitrary percentiles. The following query returns the 90-th percentile.
+HDR histograms are related to how [floating point numbers are
+represented](https://en.wikipedia.org/wiki/Double-precision_floating-point_format)
+as integers. The underlying assumption is that smaller numbers require a higher
+precision to be distinguishable (e.g. 5 ms and 6 ms are different and should be
+in different buckets) whereas larger numbers can be rounded more aggressively as
+their relative error becomes less relevant (e.g. 10000 ms and 10001 ms are
+almost the same and can reside in the same bucket).
 
-```mzsql
-SELECT bucket AS percentile
-FROM distribution
-WHERE cumulative_distribution >= 0.9
-ORDER BY cumulative_distribution
-LIMIT 1;
-```
+In the example below, to reduce the number of buckets, the values are first
+decomposed into their floating point representation (i.e., `mantissa *
```

Comment:
I came back to here when I realized that the mantissa is signed, and we don't have a separate sign bit. I think IEEE floating point has a sign bit, which was the source of my confusion. Perhaps there's a way to flag that around here?

Reply:
I also wasn't sure about this part. Removing the separate sign column makes the query simpler, so I think it's a good change. But the reference to floating point representations might be confusing. We could consider just not mentioning floats?

Fwiw, the wikipedia page for [floating point arithmetic](https://en.wikipedia.org/wiki/Floating-point_arithmetic) mostly doesn't mention a separate sign bit but uses a signed mantissa like we do here. Though your confusion makes me suspect that people are generally more familiar with the IEEE format than with the theoretical concept.

Reply:
Updated in patch 75f32afbf3839b0dc7b5f84321847c6c88161bff